client.py

```diff
@@ -0,0 +1,159 @@
+import socket, sys, logging, urllib.parse, json
+
+logging.basicConfig(filename='client_log.log',
+    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
+    level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+class client:
+
+    DEFAULT_IP = 'localhost'
+    DEFAULT_PORT = 36577
+
+    def __init__(self,ip=DEFAULT_IP,port=DEFAULT_PORT):
+        self.ip = ip
+        self.port = port
+        server_address = (ip, port)
+        logger.debug('Client instance created with IP %s port %s.' % server_address)
+
+    def __connect_socket(self):
+        # Create a TCP/IP socket
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(2)
+
+        # Connect the socket to the port where the server is listening
+        server_address = (self.ip, self.port)
+        logger.debug('connecting to %s port %s' % server_address)
+        sock.connect(server_address)
+
+        return sock
+
+    def __close_socket(self,sock):
+        logger.debug('closing socket')
+        sock.close()
+
+    def __recv(self,sock,delim='\n',recv_buffer=4096):
+        buffer = ''
+        while True:
+            data = urllib.parse.unquote_plus(sock.recv(recv_buffer).decode())
+            assert data, 'Srver disconnected while receiving.'
+            buffer += data
+            if data[-1] == delim:
+                msg = json.loads(buffer[0:-len(delim)])  # Remove delims
+                if msg['error']:
+                    raise Exception('Server Error: '+msg['response']+'\n'+msg['traceback'])
+                else:
+                    return msg['response']
+
+    def com(self,module,funcname,*args):
+        # Server always replies and always closes connection after msg
+        # assert funcname is a string, and cast varargin (cell array)
+        # to strings (use cellfun - operates on each entry of cell)
+        
+        # last input is the keep_alive; for now, functionality not
+        # included
+        assert isinstance(module,str), 'module must be a string'
+        assert isinstance(funcname,str), 'funcname must be a string'
+        
+        # Prepare both parts of message in case one errors
+        handshake = json.dumps({"name":module})
+        message = json.dumps({"function":funcname,
+            "args":args,
+            "keep_alive":False})
+
+        sock = self.__connect_socket()
+
+        resp = None
+
+        try:
+            # Send handshake
+            logger.debug('sending "%s"' % handshake)
+            sock.sendall((urllib.parse.quote_plus(handshake)+'\n').encode())
+
+            # Look for the response
+            resp = self.__recv(sock)
```

Comment:
Might be worth checking that the response is the expected response.

Reply:
You mean checking if the response is "ack"?
I added a line "assert resp == 'ack', '%s does not exist' % module". Let me know if that works.

Reply:
Yep - except I would say a more general error; "Wan't able to get an acknowledgement from the server" (your error message makes an assumption that we don't need to make)

Reply:
Got it! Just changed it.